File: packages/db/src/repository/comment.ts
import { and, asc, eq, sql } from "drizzle-orm";

import { db, dbEffect } from "../index";
import { chunk } from "../schema/chunk";
import { chunkComment } from "../schema/comment";

/**
 * SECURITY: scoped through the commented chunk's owner. `chunk_comment` has
 * a `user_id` (the comment's author) but that is not the authority here —
 * a thread on your chunk may contain other people's comments, and the right
 * to read the thread comes from owning the chunk. Without this join,
 * GET /chunks/:id/comments returned any chunk's discussion to any caller.
 */
export function listComments(chunkId: string, userId: string) {
    return dbEffect(() =>
        db
            .select({
                id: chunkComment.id,
                chunkId: chunkComment.chunkId,
                userId: chunkComment.userId,
                content: chunkComment.content,
                createdAt: chunkComment.createdAt,
                updatedAt: chunkComment.updatedAt
            })
            .from(chunkComment)
            .innerJoin(chunk, eq(chunk.id, chunkComment.chunkId))
            .where(and(eq(chunkComment.chunkId, chunkId), eq(chunk.userId, userId)))
            .orderBy(asc(chunkComment.createdAt))
    );
}

export function createComment(params: { id: string; chunkId: string; userId: string; content: string }) {
    return dbEffect(async () => {
        const [created] = await db.insert(chunkComment).values(params).returning();
        return created!;
    });
}

export function updateComment(id: string, userId: string, content: string) {
    return dbEffect(async () => {
        const [updated] = await db
            .update(chunkComment)
            .set({ content })
            .where(and(eq(chunkComment.id, id), eq(chunkComment.userId, userId)))
            .returning();
        return updated ?? null;
    });
}

export function deleteComment(id: string, userId: string) {
    return dbEffect(async () => {
        const [deleted] = await db
            .delete(chunkComment)
            .where(and(eq(chunkComment.id, id), eq(chunkComment.userId, userId)))
            .returning();
        return deleted ?? null;
    });
}

export function getCommentCount(chunkId: string) {
    return dbEffect(async () => {
        const [result] = await db
            .select({ count: sql<number>`count(*)` })
            .from(chunkComment)
            .where(eq(chunkComment.chunkId, chunkId));
        return Number(result?.count ?? 0);
    });
}

File: crates/fubbik-db/src/repo/comment.rs
//! `chunk_comment` — discussion threads on a chunk.
//!
//! # Two different owners
//!
//! This table has a `user_id`, but it is the **comment author's**, not the
//! authority for reading the thread. A thread on your chunk may contain other
//! people's comments, so:
//!
//! - **reading** a thread is gated on owning the *chunk* (join through
//!   `chunk.user_id`);
//! - **editing or deleting** a single comment is gated on having *written it*
//!   (`chunk_comment.user_id`).
//!
//! Node's `listComments` originally took only a `chunkId` and checked
//! neither, so `GET /chunks/{id}/comments` returned any chunk's discussion to
//! any authenticated caller — fixed in `bdfa534` and reproduced correctly
//! here from the start.

use fubbik_core::error::AppResult;
use sqlx::PgPool;

use crate::timestamp::UtcTimestamp;

#[derive(Debug, Clone, serde::Serialize, sqlx::FromRow, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ChunkComment {
    pub id: String,
    pub chunk_id: String,
    /// The author. See the module doc — this is not who may read the thread.
    pub user_id: String,
    pub content: String,
    #[schema(value_type = chrono::NaiveDateTime)]
    pub created_at: UtcTimestamp,
    #[schema(value_type = chrono::NaiveDateTime)]
    pub updated_at: UtcTimestamp,
}

/// The thread on a chunk, oldest first.
///
/// `, id` breaks ties on `created_at`: comments posted in the same
/// millisecond (a script, a burst) would otherwise come back in whatever
/// order the plan produced. Node orders by `created_at` alone.
pub async fn list(pool: &PgPool, chunk_id: &str, user_id: &str) -> AppResult<Vec<ChunkComment>> {
    let rows = sqlx::query_as!(
        ChunkComment,
        r#"SELECT cc.id, cc.chunk_id, cc.user_id, cc.content,
                  cc.created_at AS "created_at: UtcTimestamp",
                  cc.updated_at AS "updated_at: UtcTimestamp"
           FROM chunk_comment cc
           WHERE cc.chunk_id = $1
             AND EXISTS (SELECT 1 FROM chunk c
                         WHERE c.id = $1 AND c.user_id = $2)
           ORDER BY cc.created_at, cc.id"#,
        chunk_id,
        user_id
    )
    .fetch_all(pool)
    .await?;
    Ok(rows)
}

/// Posts a comment. Guarded on the chunk being the caller's — Node's
/// `createComment` is a bare insert whose only check is the FK, so it would
/// accept a comment on any chunk that exists.
///
/// Returns `Ok(None)` when the chunk is not the caller's.
pub async fn create(
    pool: &PgPool,
    chunk_id: &str,
    user_id: &str,
    content: &str,
) -> AppResult<Option<ChunkComment>> {
    let id = crate::new_id();
    let row = sqlx::query_as!(
        ChunkComment,
        r#"INSERT INTO chunk_comment (id, chunk_id, user_id, content)
           SELECT $1, $2, $3, $4
           WHERE EXISTS (SELECT 1 FROM chunk c WHERE c.id = $2 AND c.user_id = $3)
           RETURNING id, chunk_id, user_id, content,
                     created_at AS "created_at: UtcTimestamp",
                     updated_at AS "updated_at: UtcTimestamp""#,
        id,
        chunk_id,
        user_id,
        content
    )
    .fetch_optional(pool)
    .await?;
    Ok(row)
}

/// Edits a comment the caller **wrote**. `updated_at` is bumped explicitly:
/// Node relies on Drizzle's `$onUpdate` hook, which lives in the ORM and has
/// no database-level equivalent, so a Rust UPDATE that did not set it would
/// silently leave the column at its creation time.
pub async fn update(
    pool: &PgPool,
    id: &str,
    user_id: &str,
    content: &str,
) -> AppResult<Option<ChunkComment>> {
    let row = sqlx::query_as!(
        ChunkComment,
        r#"UPDATE chunk_comment SET content = $3, updated_at = now()
           WHERE id = $1 AND user_id = $2
           RETURNING id, chunk_id, user_id, content,
                     created_at AS "created_at: UtcTimestamp",
                     updated_at AS "updated_at: UtcTimestamp""#,
        id,
        user_id,
        content
    )
    .fetch_optional(pool)
    .await?;
    Ok(row)
}

/// Deletes a comment the caller **wrote**.
///
/// Note this means the owner of a chunk cannot delete someone else's comment
/// on it — faithful to Node, and arguably wrong for a moderation story, but
/// changing it is a product decision rather than a port decision.
pub async fn delete(pool: &PgPool, id: &str, user_id: &str) -> AppResult<Option<ChunkComment>> {
    let row = sqlx::query_as!(
        ChunkComment,
        r#"DELETE FROM chunk_comment WHERE id = $1 AND user_id = $2
           RETURNING id, chunk_id, user_id, content,
                     created_at AS "created_at: UtcTimestamp",
                     updated_at AS "updated_at: UtcTimestamp""#,
        id,
        user_id
    )
    .fetch_optional(pool)
    .await?;
    Ok(row)
}

/// Scoped through the chunk's owner, like [`list`]. Node's `getCommentCount`
/// takes only a `chunkId`, so it leaks how much discussion sits on any chunk.
pub async fn count(pool: &PgPool, chunk_id: &str, user_id: &str) -> AppResult<i64> {
    let count = sqlx::query_scalar!(
        r#"SELECT COUNT(*) AS "count!" FROM chunk_comment cc
           WHERE cc.chunk_id = $1
             AND EXISTS (SELECT 1 FROM chunk c
                         WHERE c.id = $1 AND c.user_id = $2)"#,
        chunk_id,
        user_id
    )
    .fetch_one(pool)
    .await?;
    Ok(count)
}
